akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala

```diff
@@ -92,15 +91,44 @@ private[persistence] trait Eventsourced extends Snapshotter with Stash with Stas
    * Called whenever a message replay succeeds.
    * May be implemented by subclass.
    */
-  private[persistence] def onReplaySuccess(): Unit = ()
+  private[akka] def onReplaySuccess(): Unit = ()
 
   /**
    * INTERNAL API.
-   * Called whenever a message replay fails.
-   * May be implemented by subclass.
+   * Called whenever a message replay fails. Logs the error and stops the actor.
+   * @param cause failure cause.
+   * @param event the event that was processed in `receiveRecover`, if the exception
+   *   was thrown there
+   */
+  private[akka] def onReplayFailure(cause: Throwable, event: Option[Any]): Unit = {
+    event match {
+      case Some(evt) ⇒
+        log.error(cause, "Exception in receiveRecover when replaying event type [{}] with sequence number [{}] for " +
+          "persistenceId [{}].", evt.getClass.getName, lastSequenceNr, persistenceId)
+      case None ⇒
+        log.error(cause, "Persistence failure when replaying events for persistenceId [{}]. " +
+          "Last known sequence number [{}]", persistenceId, lastSequenceNr)
+    }
+    context.stop(self)
+  }
+
+  /**
+   * Called when persist fails. Logs the error.
+   * Subclass may override to customize logging and for example send negative
+   * acknowledgment to sender.
+   *
+   * The actor is always stopped after this method has been invoked.
+   *
+   * Note that the event may or may not have been saved, depending on the type of
+   * failure.
+   *
    * @param cause failure cause.
+   * @param event the event that was to be persisted
    */
-  private[persistence] def onReplayFailure(cause: Throwable): Unit = ()
+  protected def onPersistFailure(cause: Throwable, event: Any, seqNr: Long): Unit = {
+    log.error(cause, "Failed to persist event type [{}] with sequence number [{}] for persistenceId [{}].",
+      event.getClass.getName, seqNr, persistenceId)
+  }
```

Comment:
I meant only the moving from messages to callbacks for consistency between those two?